extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -154,45 +154,35 @@ public QueryDefinition makeQueryDefinition(
 
       List<String> partitionColumnNames = new ArrayList<>();
 
-      /*
-      operatorList is a List<List<OperatorFactory>>, where each List<OperatorFactory> corresponds to the operator factories
-       to be used for a different window stage.
-
-       We iterate over operatorList, and add the definition for a window stage to QueryDefinitionBuilder.
-       */
-      for (int i = 0; i < operatorList.size(); i++) {
-        for (OperatorFactory operatorFactory : operatorList.get(i)) {
-          if (operatorFactory instanceof WindowOperatorFactory) {
-            List<String> outputColumnNames = ((WindowOperatorFactory) operatorFactory).getProcessor().getOutputColumnNames();
-
-            // Need to add column names which are present in outputColumnNames and rowSignature but not in bob,
-            // since they need to be present in the row signature for this window stage.
-            for (String columnName : outputColumnNames) {
-              int indexInRowSignature = rowSignature.indexOf(columnName);
-              if (indexInRowSignature != -1 && bob.build().indexOf(columnName) == -1) {
-                ColumnType columnType = rowSignature.getColumnType(indexInRowSignature).get();
-                bob.add(columnName, columnType);
-                log.info("Added column [%s] of type [%s] to row signature for window stage.", columnName, columnType);
-              } else {
-                throw new ISE(
-                    "Found unexpected column [%s] already present in row signature [%s].",
-                    columnName,
-                    rowSignature
-                );
-              }
+      // Iterate over the list of window stages, and add the definition for each window stage to QueryDefinitionBuilder.
+      for (int i = 0; i < stages.size(); i++) {
+        for (WindowOperatorFactory operatorFactory : stages.get(i).getWindowOperatorFactories()) {
+          // Need to add column names which are present in outputColumnNames and rowSignature but not in bob,
+          // since they need to be present in the row signature for this window stage.
+          for (String columnName : operatorFactory.getProcessor().getOutputColumnNames()) {
+            int indexInRowSignature = rowSignature.indexOf(columnName);
+            if (indexInRowSignature != -1 && bob.build().indexOf(columnName) == -1) {
+              ColumnType columnType = rowSignature.getColumnType(indexInRowSignature).get();
+              bob.add(columnName, columnType);
```

Comment:
what's this supposed to fix?

Reply:
Nothing. It's the same code, just with one less nested loop.

Previously:
```java
      for (int i = 0; i < operatorList.size(); i++) {
         for (OperatorFactory operatorFactory : operatorList.get(i)) {
           if (operatorFactory instanceof WindowOperatorFactory) {
```

Now:
```java
      for (int i = 0; i < stages.size(); i++) {
         for (WindowOperatorFactory operatorFactory : stages.get(i).getWindowOperatorFactories()) {
```

Reply:
this seems dodgy...regardless when it got here
I mean it looks up the type of an `outputColumn` from the *inputSignature* ...
it *may* be the same but I don't think there should be gurantees for that...

anyhow...this part should live as `Stage#getRowSignature` or something....

Reply:
> anyhow...this part should live as Stage#getRowSignature or something....

I tried adding WindowStage#getRowSignature. But it's not clean, we have to pass a bunch of params, and makes the code difficult to understand. So I feel we should stick with the existing code as it's cleaner and easier to follow, and we should keep WindowStage limited to the operator factory handling.

<img width="1207" alt="image" src="https://github.com/user-attachments/assets/b980b80f-accd-4d7b-a8fd-8bd412f902c5">

Reply:
I've pushed a change to create a `WindowStages` class which encapsulates `List<WindowStage>` along with some other properties. `WindowStage` wasn't self-sufficient for calculating everything for the window stage, since we rely on `window stage "x + 1"` for the shuffle spec of `window stage "x"`.

Does this approach work?

> to clean that up you need to enhance the stagebuilder

Replacing `queryDefBuilder.add(work + shuffle)` with `queryDefBuilder.add(work) + queryDefBuilder.add(shuffleNeededForNextStage)`?

I was thinking more about this. Is the end goal with the stage builder refactoring to allow each stage to do `queryDefBuilder.add(shuffleNeededForCurrentStage)` + `queryDefBuilder.add(work)`? (reversing the order, adding shuffling of current stage before work)

If my above understanding is correct, I feel this would be a bigger work item. Either way, I feel it's unrelated to the current refactoring PR, so this should be discussed separately in my opinion.